### DISCUSSION TITLE
Spatially dependent anisotropic elastic material

### DISCUSSION THREAD

**Discussion by Luca-R27** [idaholab/moose#20116] [Q&A Modules: Solid mechanics] [answered]
Dear Moose community,

I would like to simulate a material having anisotropic and spatially dependent elastic properties. In particular, let’s take the transverse isotropic case. In “ComputeElasticityTensor”, I need to specify C_ijkl = '1111, 1122, 1133, 3333, 2323’. Let’s suppose that I know how the 5 terms depend on (x, y, z) via 5 different functions of (x, y, z). 

In practice, I am looking at the generalisation of the pair “ComputeIsotropicElasticityTensor” and “ComputeVariableIsotropicElasticityTensor” to the case of anisotropic elasticity, but I could not find a “ComputeVariableElasticityTensor” that generalises “ComputeElasticityTensor”.

ComputeElasticityTensor expects C_ijkl supplied as a std::vector<double>. I could not find a way to define such vector in terms of an array of functions each returning a double. Would this do the trick, and how do I achieve this?

I also tried to define C_ijkl in a similar way as GenericFunctionRankTwoTensor is defined in the documentation:

```
[./tensor]
  type = GenericFunctionRankTwoTensor
  tensor_name = function
  tensor_functions = 'fcn_00 fcn_01 fcn_02 fcn_10 fcn_11 fcn_12 fcn_20 fcn_21 fcn_22'
  outputs = all
[../]
```

But could not find an equivalent such as "GenericFunction" array of doubles instead of a tensor.

Any help would be much appreciated.

Thank you.
Luca

**Comment by GiudGiud** [comment, marked as answer]
Hello

The format needed for this material property might not exactly exist in the framework yet, but it should be fairly easy to add.
So each component is a function, and you want to be defining a tensor, but with symmetry properties?

Guillaume

**Reply by Luca-R27** [reply]
Hello Guillaume,


Yes. For instance, quoting from the current documentation on ComputeElasticityTensor, one can define:

[./elasticity_tensor]
  #Material constants selected to match isotropic lambda and shear modulus case
  type = ComputeElasticityTensor
  C_ijkl = '1022726 113636 113636 1022726 454545'
  fill_method = axisymmetric_rz
[../]

But this does not work if I want the '1022726 113636 ... ' to be a function of (x, y, z), rather than a vector of constants. 

In addition to being able to define a tensor whose components are functions, I would need to be able to pass this tensor to the C_ijkl parameter of ComputeElasticityTensor.


Thanks,
Luca

**Reply by GiudGiud** [reply]
Yeah this isnt going work too easily from the existing code, because this is how the tensor is assigned:
```
  _elasticity_tensor[_qp] = _Cijkl;
```
so thee _Cijkl has to be already filled with Real values. 

But that doesnt make it impossible to add a ComputeFunctionElastictyTensor, it just adds an extra step.
- add a new class member _Cijkl_functions, which is a tensor of functions
- change the input parameter to a vector of functions, use the same fill method (it's templated) to fill the tensor of functions
- at evaluation time, use the tensor of functions to fill the _Cijkl tensor of Real, going component by component
- set the elasticity tensor the same way as above, using the computed _Cijkl
